### DISCUSSION TITLE
Websocket client, server간 규칙

### DISCUSSION THREAD

**Discussion by JoonhoLeecs** [ft-transcendece-team/PONG_PONG#99] [Ideas]
임시로 적은 websocket 규칙입니다. 한번 보시고 좋은 의견 알려주세요!

websocket url:
- ws://host/ws/game/<:game_type>

websocket handshaking 과정:
- code = 4001: invalid user
- code = 4009: player already have another websocket open
- code = 4000: invalid game_type

websocket content:

- text_data should have "frame", "bx", "by", "p1_pos", "p2_pos", "delta" keys
- "delta" = {"init", "p1_point", "p2_point", "winner", "p1_name", "p2_name", "p3_name", "p4_name"}: only contian one of these key when the value is changed
- frame: int / frame count since start of game(game means 1 match in 1:1, up to 3 match in trounament)
- bx: int / ball's position x coordinate 
- by: int / ball's position y coordinate
- p1_pos: int / player1's paddle position
- p2_pos: int / player2's paddle position
- delta: json / change of game state to send, {} if no change to send
- init: int / frame_count when ball will start to move
- p1_point: int / player1's point
- p2_point: int / player2's point
- winner: str "p1" or "p2" / winner of the match
- pn_name: str / player_n username

**Comment by hijkl2e** [comment]
url
1 : 1 -> one-vs-one
토너먼트 -> tournament

사용자가 게임 버튼 클릭하면
링크 이동하고
웹 소켓 잘 연결되면
빈 게임판에 loading spinner 빙글빙글 -> 이 상태가 RECRUITING
RECRUITING 상태에서 대기 제한 시간 없음, 무한 대기합니다
서버에서는 아무런 메시지도 않습니다

2명 또는 4명 모이면 PENDING 상태 - 300 프레임
RECRUITING 상태에서 PENDING 상태로 넘어갈 때 서버는 무언가 메시지를 보냅니다

**Comment by hijkl2e** [comment]
서버로부터 받는 데이터
{
    "paddle_pos": [number, number], waiting 제외하고 모든 프레임
    "ball": [number, number], waiting 제외하고 모든 프레임
    "username": [string...], 첫 PENDING
    "queue": [number...], 첫 PENDING, PENDING -> READY, 대기 중인 사용자가 나갔을 때
    "in_control": [number, number], PENDING -> READY, 플레이 중인 사용자가 나갔을 때
    "point": [number, number], PENDING -> READY, 득점이 생겼을 때
    "waiting": [number, number], 처음 연결되었을 때, 대기 인원 변경됐을 때
    "winner": number, 승자가 결정되었을 때(PLAYING -> PENDING)
    "tournament_winner": number, 토너먼트 경기가 끝났을 때
}

**Comment by hijkl2e** [comment]
프론트가 주는 데이터
{
    "target_pos": number, required
}
target_pos가 30초간 변화 없을 시 timeout

**Comment by JoonhoLeecs** [comment]
-  paddle position 상한, 하한 계산해서 서버로 보내주시는지 서버에서 상한 하한 계산하는지 궁금합니다 -> 만약 클라이언트에서 상한 하한을 계산한다면 사용자가 키를 누르더라도 상한, 하한으로 같은 맥스값 등이 들어올텐데 이런 경우에도 연결을 끊어야할까요? 

- how to notify "winner" to client when two in control disconnected? -> winner를 보내지 않는다고 들었습니다.

**Comment by hijkl2e** [comment]
프론트엔드 백엔드 모두 구현 완료되어서 닫겠습니다
